includes/payment-methods/class-upe-payment-gateway.php

```diff
@@ -0,0 +1,301 @@
+		];
+	}
+
+	/**
+	 * Returns enabled payment gateways.
+	 *
+	 * @return array
+	 */
+	public function get_enabled_payment_gateways() {
+		$enabled_gateways = [ 'card' ];
+		return $enabled_gateways;
+	}
+
+	/**
+	 * Update payment intent for completed checkout and return redirect URL for Stripe to confirm payment.
+	 *
+	 * @param int $order_id Order ID to process the payment for.
+	 *
+	 * @return array|null An array with result of payment and redirect URL, or nothing.
+	 */
+	public function process_payment( $order_id ) {
+		// TODO: Need to update Payment Intent at this point...
+		// TODO: If save_payment_method below is true, must update Payment Intent and set 'setup_future_usage' = 'off_session'.
+		$payment_intent_id = isset( $_GET['wc_payment_intent_id'] ) ? wc_clean( wp_unslash( $_GET['wc_payment_intent_id'] ) ) : ''; // phpcs:disable WordPress.Security.NonceVerification.Recommended
+		$order             = wc_get_order( $order_id );
+
+		return [
+			'result'       => 'success',
+			'redirect_url' => wp_sanitize_redirect(
+				esc_url_raw(
+					add_query_arg(
+						[
+							'order_id'            => $order_id,
+							'wc_payment_method'   => self::GATEWAY_ID,
+							'save_payment_method' => empty( $_POST[ 'wc-' . static::GATEWAY_ID . '-new-payment-method' ] ) ? 'no' : 'yes', // phpcs:ignore WordPress.Security.NonceVerification.Missing
+						],
+						$this->get_return_url( $order )
+					)
+				)
+			),
+		];
+	}
+
+	/**
+	 * Check for a redirect payment method on order received page.
+	 */
+	public function maybe_process_redirect_order() {
+		if ( ! is_order_received_page() || empty( $_GET['payment_intent_client_secret'] ) || empty( $_GET['payment_intent'] || empty( $_GET['wc_payment_method'] ) ) ) { // phpcs:ignore WordPress.Security.NonceVerification
+			return;
+		}
+
+		$payment_method = isset( $_GET['wc_payment_method'] ) ? wc_clean( wp_unslash( $_GET['wc_payment_method'] ) ) : ''; // phpcs:ignore WordPress.Security.NonceVerification
+		if ( self::GATEWAY_ID !== $payment_method ) {
+			return;
+		}
+
+		$order_id = isset( $_GET['order_id'] ) ? wc_clean( wp_unslash( $_GET['order_id'] ) ) : ''; // phpcs:ignore WordPress.Security.NonceVerification
+
+		$this->process_redirect_payment( $order_id );
+	}
+
+	/**
+	 * Processes redirect payments.
+	 *
+	 * @param int $order_id The order ID being processed.
+	 *
+	 * @throws Process_Payment_Exception When the payment intent has an error.
+	 */
+	public function process_redirect_payment( $order_id ) {
+		try {
+			$intent_id           = isset( $_GET['payment_intent'] ) ? wc_clean( wp_unslash( $_GET['payment_intent'] ) ) : ''; // phpcs:ignore WordPress.Security.NonceVerification
+			$save_payment_method = isset( $_GET['save_payment_method'] ) ? 'yes' === wc_clean( wp_unslash( $_GET['save_payment_method'] ) ) : false; // phpcs:ignore WordPress.Security.NonceVerification
+
+			if ( empty( $intent_id ) || empty( $order_id ) ) {
+				return;
+			}
```

Comment:
```suggestion
public function maybe_process_redirect_order() {
		if ( ! is_order_received_page() ) {
			return;
		}
		if ( isset( $_GET['nonce'] ) && wp_verify_nonce( sanitize_key( $_GET['nonce'] ), 'wcpay_process_redirect_order_nonce' ) ) {

			$intent_id           = isset( $_GET['payment_intent'] ) ? wc_clean( wp_unslash( $_GET['payment_intent'] ) ) : '';
			$save_payment_method = isset( $_GET['save_payment_method'] ) && 'yes' === wc_clean( wp_unslash( $_GET['save_payment_method'] ) );
			$payment_method      = isset( $_GET['wc_payment_method'] ) ? wc_clean( wp_unslash( $_GET['wc_payment_method'] ) ) : '';
			$order_id            = isset( $_GET['order_id'] ) ? wc_clean( wp_unslash( $_GET['order_id'] ) ) : '';

			if ( self::GATEWAY_ID !== $payment_method || empty( $intent_id ) ) {
				return;
			}

			$this->process_redirect_payment( $order_id, $intent_id, $save_payment_method );
		}
	}

	/**
	 * Processes redirect payments.
	 *
	 * @param int    $order_id The order ID being processed.
	 * @param string $intent_id The intent ID.
	 * @param bool   $save_payment_method True if intent should be saved in customer's account.
	 *
	 * @throws Process_Payment_Exception When the payment intent has an error.
	 */
	public function process_redirect_payment( $order_id, $intent_id, $save_payment_method ) {
		try {
```
I like to get and sanitize all the global vars at once then pass them as parameters to the other methods. Makes things a little easier with nonce verification and processing.

Reply:
Good idea: it's the season for spring cleaning and this is much tidier.